File: artsciparser.pl
# This program crawls the Arts and Science timetable pages from U of T
# It extracts the listings and places them into .csv files.
# It is considerably more tedious than crawling the Engineering timetable pages
use strict;
#use warnings;
use LWP::Simple;
use HTML::TableExtract;
use Text::CSV;
use Date::Parse;

#This function will parse the ArtSci time cells, which are formatted like 
#MW10 (mon, wed from 10-11), or MTF9-12 (mon, tue, fri from 9-12), etc
#It will also do the actual output-to-csv for each row in the table.
sub parsetime {
	#the second argument is a table row
	my @row = @{$_[1]};
	#extract the 3-letter department code (e.g. CSC, HPS, GGR, etc)
	my $dept = $row[0];
	$dept = substr $dept,0,3;
	#check first argument to determine semester (fall or winter)
	my $sem = $_[0];
	#filenames are formatted like ECE_fall.csv, CSC_winter.csv, etc
	#they are stored in either the "fall" or "winter" subfolders
	open (FH, ">>$sem/$dept"."_$sem.csv") or die "$!";
	
	#some courses have an a.m. specification. Watch out for this
	my $flag = 0;
	if (index($row[3],"a.")>0){
		$flag = 1;
	}
	
	#some cells have extra notes like (t) or (A) in them
	#they aren't needed for the purposes of this program and get in the way
	#so just strip those substrings out to make life easier
	$row[3] =~ s/\(.*//g;
	$row[3] =~ s/\ \(.*//g;
	$row[3] =~ s/<br \/>.*//g;
	$row[3] =~ s/\n.*//g;
	$row[3] =~ s/\r.*//g;
	$row[3] =~ s/a\..*//g;
	$row[3] =~ s/p.*//g;

	#make sure the row is valid to parse (course isn't TBA or canceled)
	if (length($row[3])>1 && uc($row[3]) ne "TBA" && index($row[3],',')<0){
		#extract the MTWRF alphabets (the days)
		my $dayz = $row[3];
		#strip all non-alpha characters
		$dayz =~ s/[\d-:]//g;
		#extract the hours, e.g. pull out "10-12" from MWF10-12
		my $time = $row[3];
		#strip all the leading alphabets
		$time =~ s/[M|T|W|R|F]//g;
		#convert the days into an array of letters that can be looped through
		my @days = split(undef,$dayz);
		#convert the hours into an array of numbers that can be looped through
		my @hours = split('-',$time);
		#extract the start and end times
		my $start = $hours[0];
		my $end = "";
		#integer number of hours
		if (index($row[3],':')<0){
			#course section lasts only one hour, so $end = $start+1
			$end = $start+1;
			#but if there's more than one number, $end = whatever the 2nd number is
			if ($#hours){
				$end = $hours[1];
			}
			#convert to 24-hour time for both $start and $end. 
			#Then the PHP can do proper comparisons
			if ($start<10 && $flag==0){
				if($end>$start && $end<10){
					$start+=12;
					$end+=12;
				}
			}
			if ($start<8 && $flag==0){
				if($end>$start && $end<11){
					$start+=12;
					$end+=12;
				}
			}
			if ($end<$start){
				$end+=12;
			}

			if ($start>=$end || $end>22){
				print "error ${row[0]} $start $end\n";
			}
			$start = "$start:00";
			$end = "$end:00";
		}
		#hours like x:30, x:15, etc
		else {
			my @s = strptime($start);
			#sometimes start is an integer hour, but end isn't
			if (index($hours[0],':')<0){
				$s[2] = $hours[0];
			}
			my @e = @s;
			#course section lasts only one hour, so $e[2] = $s[2]+1
			$e[2]+=1;
			#but if there's more than one number, $end = whatever the 2nd number is
			if ($#hours){
				if (index($hours[1],':')>0){
					@e = strptime($hours[1]);
				}
				#sometimes start isn't an integer hour, but end is
				else{
					$e[2] = $hours[1];
				}
			}
			#convert to 24-hour time for both $s[2] and $e[2]. 
			#Then the PHP can do proper comparisons
			if ($s[2]<10 && $flag==0){
				if($e[2]>$s[2] && $e[2]<10){
					$s[2]+=12;
					$e[2]+=12;
				}
			}
			if ($s[2]<8 && $flag==0){
				if($e[2]>$s[2] && $e[2]<11){
					$s[2]+=12;
					$e[2]+=12;
				}
			}
			if ($e[2]<$s[2]){
				$e[2]+=12;
			}
			if ($s[2]>=$e[2] || $e[2]>22){
				print "error ${row[0]} ${s[2]} ${e[2]}\n";
			}
			if (!$e[1]){
				$e[1] = "00";
			}
			if (!$s[1]){
				$s[1] = "00";
			}
			$start = "${s[2]}:${s[1]}";
			$end = "${e[2]}:${e[1]}";
		}
		#now scan each capital alphabet to determine the day of the week
		foreach my $day (@days){
			#first output the course code (e.g. ECO100Y1Y), followed by the 
			#meeting section (e.g. LEC 01), to the csv file
			print FH $row[0].$row[1],',',$row[2],',';
			#now output the day, start time, and end time to the csv file
			if ($day eq 'M'){
				print FH "Mon",',',"$start",',',"$end";
			}
			elsif ($day eq 'T'){
				print FH "Tue",',',"$start",',',"$end";
			}
			elsif ($day eq 'W'){
				print FH "Wed",',',"$start",',',"$end";
			}
			elsif ($day eq 'R'){
				print FH "Thu",',',"$start",',',"$end";
			}
			elsif ($day eq 'F'){
				print FH "Thu",',',"$start",',',"$end";
			}
			print FH ',',$row[4],"\n";
		}
	}
}

sub artsciparse {
	my $file = $_[0];
	#Extract the tables from the HTML page to copy into memory
	my $raw_html = do {
		open my $in, '<', "$file"
		    or die "Can't open infile: $!\n";
		local $/ = undef;
		<$in>;
	};

	#We only want these four columns from the page. Details like location and prof
	#are less important. The main focus is timing.
	my $te = new HTML::TableExtract(headers => [qw(Course SC Meeting\nSection Time Location)]);
	#this creates a single table, unlike the multiple tables created 
	#when parsing engineering pages, because there's only ONE unique
	#header on the page this time.
	#the tables are stored in memory first before being saved to csv
	$te->parse($raw_html);

	# Examine the resulting table
	my $ts = ($te->tables)[0];
	#to avoid reference errors, copy the table into this local variable
	my @table = $ts;
	if ($ts){
		@table = $ts->rows;
	}
	#count number of rows in the table
	my $numrows = $#table+1;
	
	#loop through the table and analyze it
	for (my $i = 0; $i<$numrows; $i++) {
		my $name = $table[$i][0];
		my $sem = $table[$i][1];
		$table[$i][2]= substr $table[$i][2],0,5;
		#strip out info like notes from the location cell
		$table[$i][4] =~ s/\r.*//g;
		$table[$i][4] =~ s/\n.*//g;
		my $sect = $table[$i][2];
		#if the name/section is blank, then we want to copy the value from above
		#would prefer to check if it's null/undef/zero 
		#but those comparisons always seem to fail
		if (length($name)<7){
			#copy the code and semester (in artsci, the F/S/Y part is separate)
			$table[$i][0] = $table[$i-1][0];
			$table[$i][1] = $table[$i-1][1];
			#if blank, copy the section (LEC 01, PRA 02, etc), from the
			#above row into the current row.
			if (length($sect)<4) {
				$table[$i][2] = $table[$i-1][2];
			}
		}
		
	}
	#finally, loop through the reformatted table and output each row to CSV
	foreach my $row (@table) {
		if (@$row[1] eq 'F' || @$row[1] eq 'Y'){
			#got to account for some stupid comma cases (one cell had the value "T3, R12")
			if(index(@$row[3],',')>0){
				my @arr = split(/, /, @$row[3]);
				foreach my $x (@arr){
					$x =~ s/\n//g;
					@$row[3] = $x;
					parsetime("fall",\@$row);
				}
			}
			else{
				#strip out trailing whitespace in case some idiot put it there
				@$row[3] =~ s/\ //g;
				parsetime("fall",\@$row);
			}
		}
		if (@$row[1] eq 'S' || @$row[1] eq 'Y'){
			#got to account for some stupid comma cases (one cell had the value "T3, R12")
			if(index(@$row[3],',')>0){
				my @arr = split(/, /, @$row[3]);
				foreach my $x (@arr){
					$x =~ s/\n//g;
					@$row[3] = $x;
					parsetime("winter",\@$row)
				}
			}
			else {
				#strip out trailing whitespace in case some idiot put it there
				@$row[3] =~ s/\ //g;
				parsetime("winter",\@$row);
			}
		}
	}
}

#parse all the downloaded files from the artsci page
my @files = <artsci/*.html>;
foreach my $file (@files) {
	artsciparse($file);
}

File: artsciparser.py
import urllib2, os
from BeautifulSoup import *
from time import *
import re

#grab the links to the individual sites from the main site
def getlinks():
	mainsite=urllib2.urlopen("http://www.artsandscience.utoronto.ca/ofr/timetable/winter/sponsors.htm")
	soup=BeautifulSoup(mainsite.read())
	links = []
	for link in soup.findAll('a'):
		if link.has_key('href') and link['href'][0:4]!="http" and link['href'][0:4]!="mail":
			links.append("http://www.artsandscience.utoronto.ca/ofr/timetable/winter/"+link['href'])
	return links

#convert time to a readable format
def converttime(time):
	flag = 0
	if time.find("a.")>0:
		flag = 1;
	s=0
	e=0
	start = time.split('-')[0]
	try:
		end = time.split('-')[1]
	except:
		if start==start.split(':')[0]:
			end = str(int(start)+1)
		else:
			s = strptime(start,"%I:%M")
			end = str(s.tm_hour+1)+':'+str(s.tm_min)
	#get s and e as integer hours
	if start==start.split(':')[0]:
		s = int(start)
	else:
		s = strptime(start,"%I:%M").tm_hour
	if end==end.split(':')[0]:
		e = int(end)
	else:
		e = strptime(end,"%I:%M").tm_hour
	#convert to 24 hours
	if s<10 and flag==0:
		if e>s and e<10:
			s = s+12
			e = e+12
	if s<8 and flag==0:
		if e>s and e<11:
			s = s+12
			e = e+12
	if e<s:
		e = e+12
	if s>=e or e>22:
		print "error"+s+e
	#now convert back to string
	if start==start.split(':')[0]:
		start = str(s)+":00"
	else:
		start = str(s)+":"+str(strptime(start,"%I:%M").tm_min)
	if end==end.split(':')[0]:
		end = str(e)+":00"
	else:
		end = str(e)+":"+str(strptime(end,"%I:%M").tm_min)
	
	return {0:start,1:end}

#output row to csv, very important to convert the time part first
def outputtocsv(row,sem):
	#day codes
	daymapping = {'M':"Mon",'T':"Tue",'W':"Wed",'R':"Thu",'F':"Fri"}
	#some courses have an a.m. specification. Watch out for this
	flag = 0
	row[2] = row[2].split('\r')[0]
	row[2] = row[2].split('\n')[0]
	row[2] = row[2].split('(')[0]
	#no commas - easier scenario by a long shot
	if row[2]==row[2].split(',')[0]:
		row[2] = row[2].split(' ')[0]
		#strip out anything that's not M,T,W,R,or F
		days = re.sub('[^M|T|W|R|F]+', '', row[2])
		#strip out the opposite of above
		time = re.sub('[M|T|W|R|F]+', '', row[2])
		for day in list(days):
			day = daymapping[day]
			time1 = converttime(time)
			start = time1[0]
			end = time1[1]
			line = row[0]+','+row[1]+','+day+','+start+','+end+','+row[3]
			#get department code, e.g. it is CSC if the course is CSC180H1F
			dept = row[0][0:3]
			#open (or create if not found) a csv file to store the timings
			#filenames are formatted like ECE_fall.csv, CSC_winter.csv, etc
			#they are stored in either the "fall" or "winter" subfolders
			f = open(sem+"/"+dept+'_'+sem+".csv",'a')
			f.write(line+'\n')
			f.close()
	#some idiot put the cell as something like M10,T2 
	#instead of having 2 separate rows (probably to make life hard)
	else: 
		#split on the comma
		slots = row[2].split(', ')
		for slot in slots:
			days = re.sub('[^M|T|W|R|F]+', '', slot)
			time = re.sub('[M|T|W|R|F]+', '', slot)
			for day in list(days):
				day = daymapping[day]
				time1 = converttime(time)
				start = time1[0]
				end = time1[1]
				line = row[0]+','+row[1]+','+day+','+start+','+end+','+row[3]
				#get department code, e.g. it is CSC if the course is CSC180H1F
				dept = row[0][0:3]
				#open (or create if not found) a csv file to store the timings
				#filenames are formatted like ECE_fall.csv, CSC_winter.csv, etc
				#they are stored in either the "fall" or "winter" subfolders
				f = open(sem+"/"+dept+'_'+sem+".csv",'a')
				f.write(line+'\n')
				f.close()

#process each site and grab the tables
def artsciparser(link):
	#grab the URL remotely and start parsing it
	site=urllib2.urlopen(link)
	soup=BeautifulSoup(site.read())
	#Examine all HTML tables in the site
	for table in soup.findAll('table'):
		mytable = []
		rows = table.findAll('tr')
		#convert to 2D array first
		for i in range (3,len(rows)):
			cols = rows[i].findAll('td')
			try:
				check = cols[5].find(text=True)
				if check[0:3].upper()!="TBA":
					new = []
					for field in cols:						
						#replace all &nbsps with blanks - easier to check
						if field.find(text=True)=="&nbsp;":
							new.append("")
						else:
							new.append(field.find(text=True))
					new[0] = new[0]+new[1]
					del new[4]
					del new[2]
					del new[1]
					mytable.append(new)
			except:
				pass
		#clear up blank cells
		for i in range (0,len(mytable)):
			row = mytable[i]
			try:
				#strip out info like notes from the location cell
				row[3] = row[3].split('\r')[0]
				row[3] = row[3].split('\n')[0]
			except:
				pass
			#if the name/section is blank, then we want to copy the value from above
			if row[0]=="":
				row[0] = mytable[i-1][0]
				if row[1]=="":
					row[1] = mytable[i-1][1]
			#finally convert to csv
			try:
				if row[2]!="":
					if row[0][-1]=='Y' or row[0][-1]=='F':
						outputtocsv(row,"fall")
					if row[0][-1]=='Y' or row[0][-1]=='S':
						outputtocsv(row,"winter")
			except:
				print link

